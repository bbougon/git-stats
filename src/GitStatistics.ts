import { MergeEventRepository, mergeEventsStatistics } from "./merge-events/MergeEvent.js";
import { RequestParameters } from "../index.js";

type GitEvent = object;
type GitEventsStatisticsResult = object;
type Period = { start: Date; end: Date };

interface GitStatistics {
  readonly period: Period;

  result: () => GitEventsStatisticsResult;

  sortedEvents: () => GitEvent[];
}

type StatisticsAggregate = { [key: string]: GitStatistics };
const gitStatistics = (
  requestParameter: RequestParameters,
  repository: MergeEventRepository
): Promise<StatisticsAggregate> => {
  return mergeEventsStatistics(repository, requestParameter).then((statistics) =>
    Promise.resolve({ mergedEvents: statistics })
  );
};
type Unit = string | "Week" | "Month";

class Dimension {
  private constructor(public readonly unit: Unit, public readonly index: number, public mr: number) {}

  static create(unit: Unit, index: number) {
    return new Dimension(unit, index, 1);
  }

  increase() {
    this.mr = this.mr + 1;
  }

  static empty(unit: Unit, index: number) {
    return new Dimension(unit, index, 0);
  }
}

export { Dimension };
export { Unit };
export { gitStatistics };
export { StatisticsAggregate };
export { GitStatistics };
export { Period };
export { GitEventsStatisticsResult };
export { GitEvent };
