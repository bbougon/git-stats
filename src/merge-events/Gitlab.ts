import { RequestParameters } from "../../index.js";

export type MergeRequestsStatsParameters = RequestParameters & {
  projectId: number;
};
