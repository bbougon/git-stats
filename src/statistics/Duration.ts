type Duration = {
  months: number;
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
};
export default Duration;
