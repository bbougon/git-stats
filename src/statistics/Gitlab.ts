import { RequestParameters } from "../../index.js";

export type GitlabEventParameters = RequestParameters & {
  projectId: number;
};
