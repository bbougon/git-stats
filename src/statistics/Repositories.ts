import { MergeEvent } from "./merge-events/MergeEvent.js";
import { IssueEvent } from "./issues/Issues.js";
import { EventRepository } from "./EventRepository.js";

abstract class Repositories {
  private static repositories: Repositories;

  static initialize(repositories: Repositories) {
    Repositories.repositories = repositories;
  }

  static mergeEvent(): EventRepository<MergeEvent> {
    return Repositories.repositories.getMergeEventRepository();
  }

  static issueEvent(): EventRepository<IssueEvent> {
    return Repositories.repositories.getIssueEventRepository();
  }

  protected abstract getMergeEventRepository(): EventRepository<MergeEvent>;

  protected abstract getIssueEventRepository(): EventRepository<IssueEvent>;
}

export { Repositories };
