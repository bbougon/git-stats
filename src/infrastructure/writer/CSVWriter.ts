import { Writer } from "../../../index.js";
import fs from "fs";
import { stringify } from "csv-stringify/sync";
import moment from "moment";
import { StatisticsAggregate } from "../../statistics/GitStatistics.js";

export class CSVWriter implements Writer {
  private HEADER = [
    { key: "eventType", header: "Event" },
    { key: "project", header: "Project" },
    { key: "id", header: "id" },
    { key: "createdAt", header: "Created At" },
    { key: "mergedAt", header: "Merged At" },
    { key: "closedAt", header: "Closed At" },
  ];
  constructor(private readonly filePath: string) {}

  write(stats: StatisticsAggregate): void {
    try {
      const reportFilePath = `${this.filePath}/report`;
      if (!fs.existsSync(reportFilePath)) {
        fs.mkdirSync(reportFilePath);
      }
      const input = Object.entries(stats).flatMap(([key, value]) =>
        value.sortedEvents().map((event) => ({ eventType: key, ...event }))
      );
      const stringifier = stringify(input, {
        header: true,
        columns: this.HEADER,
        delimiter: ";",
        cast: {
          date: function (value: Date) {
            return { value: moment(value).toISOString(), quote: false };
          },
        },
      });
      fs.writeFileSync(`${reportFilePath}/report.csv`, stringifier);
    } catch (e) {
      console.log(e);
    }
  }
}
