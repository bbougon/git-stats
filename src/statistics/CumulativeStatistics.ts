import { GitEvent, GitEventsStatisticsResult, GitStatistics, Period } from "./Statistics.js";
import { Unit } from "./GitStatistics.js";
import { eachMonthOfInterval, eachWeekOfInterval, endOfMonth, endOfWeek, getMonth, getWeek, getYear } from "date-fns";

interface CumulativeStatistic {
  readonly index: number;
  readonly period: Period;
  readonly opened: number;
  readonly closed: number;
  trend: number;
}

type CumulativeStatisticsResult = GitEventsStatisticsResult & {
  cumulativeResults: Map<Unit, CumulativeStatistic[]>;
};

class CumulativeStatistics implements GitStatistics {
  constructor(
    readonly events: GitEvent[],
    readonly period: Period,
    private readonly eventDate: (event: GitEvent) => Period
  ) {}

  result(): CumulativeStatisticsResult {
    const stats = new Map<Unit, CumulativeStatistic[]>();

    const weeksCumulativeStatistics = this.cumulativeStatistics(
      eachWeekOfInterval({
        end: this.period.end,
        start: this.period.start,
      }),
      (date, compareTo) => getWeek(date) === getWeek(compareTo) && getYear(date) === getYear(compareTo),
      (week) => endOfWeek(week),
      (date) => getWeek(date)
    );
    const monthsCumulativeStatistics = this.cumulativeStatistics(
      eachMonthOfInterval({
        end: this.period.end,
        start: this.period.start,
      }),
      (date, compareTo) => getMonth(date) === getMonth(compareTo) && getYear(date) === getYear(compareTo),
      (month) => endOfMonth(month),
      (date) => getMonth(date)
    );

    TrendCalculator.calculate(weeksCumulativeStatistics);
    TrendCalculator.calculate(monthsCumulativeStatistics);
    stats.set("Week", weeksCumulativeStatistics);
    stats.set("Month", monthsCumulativeStatistics);

    return { cumulativeResults: stats };
  }

  private cumulativeStatistics(
    periodsInInterval: Date[],
    comparingDate: (date: Date, compareTo: Date) => boolean,
    endOfPeriod: (date: Date) => Date,
    indexNumber: (date: Date) => number
  ): CumulativeStatistic[] {
    const result: CumulativeStatistic[] = [];
    periodsInInterval.forEach((date, index) => {
      const opened = this.events.filter((event) => {
        return comparingDate(date, this.eventDate(event).start);
      }).length;
      const closed = this.events.filter((event) => {
        return comparingDate(date, this.eventDate(event).end);
      }).length;
      if (result.length === 0) {
        const cumulativeStatistics = {
          index: indexNumber(date),
          period: { end: endOfPeriod(date), start: date },
          opened,
          closed,
          trend: 0,
        };
        result.push(cumulativeStatistics);
      } else {
        const cumulativeResults = result[index - 1];
        const cumulativeStatistics = {
          index: indexNumber(date),
          period: {
            end: endOfPeriod(date),
            start: date,
          },
          opened: cumulativeResults.opened + opened,
          closed: cumulativeResults.closed + closed,
          trend: 0,
        };
        result.push(cumulativeStatistics);
      }
    });
    return result;
  }
}

class TrendCalculator {
  static calculate(cumulativeStatistics: CumulativeStatistic[]) {
    const { sumOfIndexesByOpened, sumOpened, sumIndex, sumOfSquareIndex } = cumulativeStatistics.reduce(
      (accumulator, currentValue, currentIndex) => {
        return {
          sumOfIndexesByOpened: accumulator.sumOfIndexesByOpened + (currentIndex + 1) * currentValue.opened,
          sumOpened: accumulator.sumOpened + currentValue.opened,
          sumIndex: accumulator.sumIndex + (currentIndex + 1),
          sumOfSquareIndex: accumulator.sumOfSquareIndex + Math.pow(currentIndex + 1, 2),
        };
      },
      { sumOfIndexesByOpened: 0, sumOpened: 0, sumIndex: 0, sumOfSquareIndex: 0 }
    );
    const slope =
      (cumulativeStatistics.length * sumOfIndexesByOpened - sumOpened * sumIndex) /
        (cumulativeStatistics.length * sumOfSquareIndex - Math.pow(sumIndex, 2)) || 0;
    const yIntercept = parseFloat(((sumOpened - slope * sumIndex) / cumulativeStatistics.length).toFixed(1));
    cumulativeStatistics.forEach(
      (cumulativeStatistic, index) => (cumulativeStatistic.trend = slope * (index + 1) + yIntercept)
    );
  }
}

export { CumulativeStatistic, CumulativeStatisticsResult, TrendCalculator, CumulativeStatistics };
