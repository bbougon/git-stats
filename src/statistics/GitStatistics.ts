import { MergeEventRepository, mergeEventsStatistics } from "./merge-events/MergeEvent.js";
import { differenceInHours, getMonth, getWeek } from "date-fns";
import { RequestParameters } from "../../index.js";
import moment from "moment/moment.js";
import Duration from "./Duration.js";

type GitEvent = object;
type GitEventsStatisticsResult = object;
type Period = { start: Date; end: Date };

interface GitStatistics {
  readonly period: Period;

  result: () => GitEventsStatisticsResult;

  sortedEvents: () => GitEvent[];
}

type StatisticsAggregate = { [key: string]: GitStatistics };

const gitStatistics = (
  requestParameter: RequestParameters,
  repository: MergeEventRepository
): Promise<StatisticsAggregate> => {
  return mergeEventsStatistics(repository, requestParameter).then((statistics) =>
    Promise.resolve({ mergedEvents: statistics })
  );
};
type Unit = string | "Week" | "Month";
export type Year = number;
export class StatisticFlow {
  public readonly events: Period[] = [];

  constructor(period: Period | undefined, public readonly index: number) {
    if (period !== undefined) {
      this.events.push(period);
    }
  }

  total() {
    return this.events.length;
  }

  addEvent(period: Period) {
    this.events.push(period);
  }

  static month(period: Period): StatisticFlow {
    return new StatisticFlow(period, getMonth(period.end));
  }

  static week(period: Period): StatisticFlow {
    return new StatisticFlow(period, getWeek(period.end));
  }

  static empty(index: number) {
    return new StatisticFlow(undefined, index);
  }

  average(): Duration {
    const hoursSpent = this.events.reduce(
      (accumulator, currentValue) => accumulator + differenceInHours(currentValue.end, currentValue.start),
      0
    );
    const duration = moment.duration(hoursSpent / this.events.length, "hours");
    return {
      months: parseInt(duration.months().toFixed()),
      days: parseInt(duration.days().toFixed()),
      hours: parseInt(duration.hours().toFixed()),
      minutes: parseInt(duration.minutes().toFixed()),
      seconds: parseInt(duration.seconds().toFixed()),
    };
  }

  median(): Duration {
    const hoursSeries = this.events.map((period) => ({ hours: differenceInHours(period.end, period.start) }));
    hoursSeries.sort((current, next) => (current.hours > next.hours ? 1 : -1));
    const seriesIndex = (hoursSeries.length + 1) / 2 - 1;
    const isHourSeriesEvent = (hoursSeries.length + 1) % 2 === 0;
    let duration = {} as moment.Duration;
    if (isHourSeriesEvent) {
      duration = moment.duration(hoursSeries[seriesIndex].hours, "hours");
    } else {
      const hours = (hoursSeries[Math.floor(seriesIndex)].hours + hoursSeries[Math.ceil(seriesIndex)].hours) / 2;
      duration = moment.duration(hours, "hours");
    }
    return {
      months: parseInt(duration.months().toFixed()),
      days: parseInt(duration.days().toFixed()),
      hours: parseInt(duration.hours().toFixed()),
      minutes: parseInt(duration.minutes().toFixed()),
      seconds: parseInt(duration.seconds().toFixed()),
    };
  }
}
export const gitEventsByPeriod = (
  gitEventStatistics: GitStatistics,
  eventDate: (event: GitEvent) => Period
): Map<Year, { [p: Unit]: StatisticFlow[] }[]> => {
  const stats: Map<Year, { [key: Unit]: StatisticFlow[] }[]> = new Map<Year, { [key: Unit]: StatisticFlow[] }[]>();
  gitEventStatistics
    .sortedEvents()
    .filter((event) => eventDate(event).end !== null)
    .forEach((event) => {
      const period = eventDate(event);
      const year = period.end.getFullYear();
      const monthFlow = StatisticFlow.month(period);
      const weekFlow = StatisticFlow.week(period) as StatisticFlow;
      const yearStats = stats.get(year);

      function addFlow(flows: StatisticFlow[], flow: StatisticFlow) {
        const existingFlow = flows.filter((existingFlow: StatisticFlow) => existingFlow.index === flow.index);
        if (existingFlow.length === 0) {
          flows.push(flow);
        }
        existingFlow.forEach((flow: StatisticFlow) => {
          flow.addEvent(period);
        });
      }

      if (yearStats === undefined) {
        stats.set(year, [{ Month: [monthFlow] }, { Week: [weekFlow] }] as { [key: Unit]: StatisticFlow[] }[]);
      } else {
        yearStats.forEach((period) => {
          Object.entries(period).forEach(([key, flows]) => {
            if (key === "Month") {
              addFlow(flows, monthFlow);
            }
            if (key === "Week") {
              addFlow(flows, weekFlow);
            }
          });
        });
      }
    });
  return fillEmptyPeriodsAndSortChronologically(stats, gitEventStatistics.period);
};

type PeriodIndexes = {
  firstPeriodIndex: number;
  lastPeriodIndex: number;
};

abstract class PeriodIndexesBuilder {
  static periodIndexes = new Map<Unit, PeriodIndexesBuilder>([
    [
      "Month",
      new (class extends PeriodIndexesBuilder {
        protected build(currentYear: Year, period: Period): PeriodIndexes {
          const periodIndex = this.initialize(period, getMonth);
          if (currentYear !== period.end.getFullYear()) {
            periodIndex.lastPeriodIndex = 12;
          }
          if (currentYear !== period.start.getFullYear()) {
            periodIndex.firstPeriodIndex = 0;
          }
          return periodIndex;
        }
      })(),
    ],
    [
      "Week",
      new (class extends PeriodIndexesBuilder {
        protected build(currentYear: Year, period: Period): PeriodIndexes {
          const periodIndex = this.initialize(period, getWeek);
          if (currentYear !== period.end.getFullYear()) {
            periodIndex.lastPeriodIndex = getWeek(new Date(currentYear, 11, 31)) + 1;
          }
          if (currentYear !== period.start.getFullYear()) {
            periodIndex.firstPeriodIndex = 1;
          }
          return periodIndex;
        }
      })(),
    ],
  ]);

  static for(currentYear: Year, period: Period, unit: Unit): PeriodIndexes {
    return this.periodIndexes.get(unit).build(currentYear, period);
  }

  protected abstract build(year: Year, period: Period): PeriodIndexes;
  protected initialize(period: Period, getDate: (date: Date) => number): PeriodIndexes {
    return { firstPeriodIndex: getDate(period.start), lastPeriodIndex: getDate(period.end) + 1 };
  }
}

const fillEmptyPeriodsAndSortChronologically = (
  stats: Map<Year, { [key: Unit]: StatisticFlow[] }[]>,
  period: Period
): Map<Year, { [key: Unit]: StatisticFlow[] }[]> => {
  const completeStatistics = stats;

  function fillEmptyPeriodsInInterval(_stat: { [p: Unit]: StatisticFlow[] }, year: Year) {
    Object.entries(_stat).forEach(([unit, statisticFlows]) => {
      const periodIndexes = PeriodIndexesBuilder.for(year, period, unit);
      while (periodIndexes.firstPeriodIndex < periodIndexes.lastPeriodIndex) {
        const currentPeriodIndex = periodIndexes.firstPeriodIndex;
        if (statisticFlows.find((currentStat) => currentStat.index === currentPeriodIndex) === undefined) {
          statisticFlows.push(StatisticFlow.empty(currentPeriodIndex));
        }
        periodIndexes.firstPeriodIndex++;
      }
      statisticFlows.sort((current, next) => (current.index > next.index ? 1 : -1));
    });
  }

  completeStatistics.forEach((flows, year) => {
    flows.forEach((period) => {
      fillEmptyPeriodsInInterval(period, year);
    });
  });
  return completeStatistics;
};
export { Unit, StatisticsAggregate, GitStatistics, Period, GitEventsStatisticsResult, GitEvent };
export { gitStatistics };
