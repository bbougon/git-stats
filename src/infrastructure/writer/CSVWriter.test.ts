import { IssueEventBuilder, MergeEventBuilderForMR } from "../../__tests__/builder.js";
import { mkdtemp } from "node:fs/promises";
import path from "path";
import os from "os";
import fs from "fs";
import { parseISO } from "date-fns";
import { CSVWriter } from "./CSVWriter.js";
import {
  GitlabMemoryRepositories,
  IssueEventsMemoryRepository,
  MergeRequestMemoryRepository,
} from "../../__tests__/MemoryRepositories";
import { gitStatistics } from "../../statistics/GitStatistics";
import { GitlabEventParameters } from "../../statistics/Gitlab";
import { Repositories } from "../../statistics/Repositories";

describe("CSV writer", () => {
  beforeEach(() => {
    new GitlabMemoryRepositories();
  });

  test("should generate a CSV report file with all merge events and issue events", async () => {
    const mergeRequestBuilder = new MergeEventBuilderForMR(1);
    const firstMergeRequest = mergeRequestBuilder
      .id(1)
      .createdAt(parseISO("2022-02-11T12:37:22+01:00"))
      .mergedAt(parseISO("2022-02-14T11:53:17+01:00"))
      .build();
    const secondMergeRequest = mergeRequestBuilder
      .id(2)
      .createdAt(parseISO("2022-02-12T13:22:54+01:00"))
      .notYetMerged()
      .build();
    const thirdMergeRequest = mergeRequestBuilder
      .id(3)
      .createdAt(parseISO("2022-02-13T09:17:34+01:00"))
      .closedAt(parseISO("2022-02-16T16:44:22+01:00"))
      .build();
    (Repositories.mergeEvent() as MergeRequestMemoryRepository).persistAll([
      firstMergeRequest,
      secondMergeRequest,
      thirdMergeRequest,
    ]);
    const firstIssueEvent = new IssueEventBuilder(1).withId(1).createdAt(parseISO("2022-02-11T12:37:22+01:00")).build();
    const secondIssueEvent = new IssueEventBuilder(1)
      .withId(2)
      .createdAt(parseISO("2022-02-11T12:37:22+01:00"))
      .closedAt(parseISO("2022-02-14T11:53:17+01:00"))
      .build();
    (Repositories.issueEvent() as IssueEventsMemoryRepository).persistAll([firstIssueEvent, secondIssueEvent]);
    const tempDirectory = await mkdtemp(path.join(os.tmpdir(), "report-"));
    const fromDate = parseISO("2022-02-11T00:00:00");
    const toDate = parseISO("2022-02-17T00:00:00");

    const stats = await gitStatistics({
      projectId: 1,
      fromDate: fromDate,
      toDate: toDate,
    } as GitlabEventParameters);
    new CSVWriter(tempDirectory).write(stats);

    expect(fs.readFileSync(`${tempDirectory}/report/report.csv`, "utf8")).toMatchSnapshot();
  });
});
