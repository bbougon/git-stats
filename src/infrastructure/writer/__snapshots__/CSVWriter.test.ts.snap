// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSV writer should generate a CSV report file with all merge events and issue events 1`] = `
"Event;Project;id;Created At;Merged At;Closed At
mergeEvents;1;1;2022-02-11T11:37:22.000Z;2022-02-14T10:53:17.000Z;
mergeEvents;1;2;2022-02-12T12:22:54.000Z;;
mergeEvents;1;3;2022-02-13T08:17:34.000Z;;2022-02-16T15:44:22.000Z
issues;1;1;2022-02-11T11:37:22.000Z;;
issues;1;2;2022-02-11T11:37:22.000Z;;2022-02-14T10:53:17.000Z
"
`;
